migration/scale.go

```diff
@@ -109,43 +111,68 @@ func (s scaleUp) deployments(ctx context.Context, client kubernetes.Interface) e
 func (s scaleDown) statefulSets(ctx context.Context, client kubernetes.Interface) error {
 	sets, err := client.AppsV1().StatefulSets(s.namespace).List(ctx, metav1.ListOptions{})
 	if err != nil {
-		return nil
+		return err
 	}
+
 	for _, set := range sets.Items {
-		for _, vol := range set.Spec.Template.Spec.Volumes {
-			if vol.PersistentVolumeClaim == nil {
-				continue
-			}
-			if vol.PersistentVolumeClaim.ClaimName != s.pvc {
-				continue
-			}
-			// ensure that we have not processed it before already
-			if _, ok := set.Annotations[pvcMigration]; ok {
-				continue
-			}
+		if !s.shouldScale(set) {
+			continue
+		}
 
-			// has PVC, scale down
-			if set.Spec.Replicas == nil {
-				addAnnotation(&set, oldReplicaCount, "-1")
-			} else {
-				addAnnotation(&set, oldReplicaCount, strconv.Itoa(int(*set.Spec.Replicas)))
-			}
-			addAnnotation(&set, pvcMigration, s.pvc)
-			set.Spec.Replicas = new(int32) // default 0
+		// has PVC, scale down
+		if set.Spec.Replicas == nil {
+			addAnnotation(&set, oldReplicaCount, "-1")
+		} else {
+			addAnnotation(&set, oldReplicaCount, strconv.Itoa(int(*set.Spec.Replicas)))
+		}
+		addAnnotation(&set, pvcMigration, s.pvc)
+		set.Spec.Replicas = new(int32) // default 0
 
-			_, err := client.AppsV1().StatefulSets(s.namespace).Update(ctx, &set, metav1.UpdateOptions{})
-			if err != nil {
-				return errors.Wrapf(err, "could not update deployment %v", set.Name)
-			}
+		_, err := client.AppsV1().StatefulSets(s.namespace).Update(ctx, &set, metav1.UpdateOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "could not update statefulSet %v", set.Name)
+		}
 
-			log.Debugf("%v/%v: scaled down statefulSet %v", s.namespace, s.pvc, set.Name)
+		log.Debugf("%v/%v: scaled down statefulSet %v", s.namespace, s.pvc, set.Name)
 
-			break
-		}
+		break
```

Comment:
I just realised: even if you have an RWX storage solution, you'll never be able to share storage between separate statefulsets or deployments? Which in turn explains why it's safe to `break` here.

Reply:
Deployments should not matter since they are handled elsewhere, but I guess there is nothing stopping you from mounting an RWX volume in multiple StatefulSets. So in this case, we would miss scaling down the second one.

Reply:
> I'm actually not sure how this works; if you have an RWX volume, do you specify the same PVC everywhere?

Yes, in the end the same PVC will just be mounted by multiple pods (which is of course possible with RWX), so it should not really matter which controller creates the pod. So I think we should probably remove the break, even though this is a pretty niche use case.